torch/_inductor/lowering.py

```diff
@@ -6219,10 +6220,199 @@ def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs):
             msg = f"{msg} Found from : \n {stack_trace}"
         V.graph.disable_cudagraphs_reason = msg
 
+    fx_all_inputs = V.graph.current_node.args[-2] + V.graph.current_node.args[-1]  # type: ignore[operator]
+    fake_all_inputs = [x.meta["val"] for x in fx_all_inputs]  # type: ignore[union-attr]
+
+    for subgraph in (cond_fn, body_fn):
+        if subgraph.graph is None:
+            # create and lower subgraphs
+            subgraph.graph = V.graph.make_subgraph(
+                gm=subgraph.graph_module,
+                example_inputs=fx_all_inputs,  # type: ignore[arg-type]
+                subgraph_name=subgraph.name,
+            )
+            with V.set_graph_handler(subgraph.graph):
+                subgraph.graph.run(*fake_all_inputs)
+
     result = ir.WhileLoop.create(cond_fn, body_fn, carried_inputs, additional_inputs)
     return list(map(TensorBox.create, result))
 
 
+# This is a spcialized version of select because we're certain that the idx is valid
+@register_lowering(torch.ops._scan_helper.unsafe_select, type_promotion_kind=None)
+def _(dst, dim, idx):
+    return squeeze(slice_(dst, dim, idx, idx + 1, step=1, clamp=False), dim)
+
+
+@register_lowering(scan_op)
+def scan(combine_subgraph, init, xs, dim, reverse, additional_inputs):
+    from torch._dynamo.source import GlobalSource
+    from torch._higher_order_ops.scan import _extract_carry_and_out, stack_y
+    from torch._inductor.pattern_matcher import fwd_only
+    from torch.fx.experimental.symbolic_shapes import DimDynamic
+
+    if torch._inductor.config.cpp_wrapper:
+        raise NotImplementedError("scan is not supported with cpp_wrapper yet.")
+
+    num_init_leaves = len(init)
+    specialized_dim = int(dim)
+
+    with V.graph.fake_mode:
+        # We choose tensor box to be the single source of truth of creating fake.
+        # tensors for subgraph.
+        # It's not reliable to get fake tensor from the "val" meta of fx node.
+        # For example: the View ir node in inductor will create a new TensorBox
+        # whose size is exactly the viewed shape but in fake tensor
```

Comment:
I don't understand why this is needed? Why won't the size be the same shape?

Reply:
The size is the same semantically but may use different symbols. something like `s6//s0` vs `s4//s0` when we torch.compile(dynamic=True), which assigns two symbols to the same input size of _input and _target in the below code:
```python
chunks = _input.shape[0] // CHUNK_SIZE
_input_chunks = _input.view(CHUNK_SIZE, chunks, *_input.shape[1:])
target_chunks = target.view(CHUNK_SIZE, chunks, *target.shape[1:])
```
target_chunks is `(s0, s6//s0, s3)` in fx_node but is `(s0, s4//s0, s3)` in its inductor IR tensor box.get_size(), due to the logic of lowering View in inductor (which just overrides with new size). This fails the shape check of while loop in L6639-6640 in ir.py: https://github.com/pytorch/pytorch/pull/135603/files/f85066f615fb0a95c849ff99d61297daa2d15006#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acL6639-L6640. 



In fact, I tried to change the `==` into statically_known_equals but it fails at the stride check like below:
```python
# True
V.graph.sizevars.statically_known_equals(s6//s0, s4//s0)
# False
V.graph.sizevars.statically_known_equals(s0*s6//s0, s0*s4//s0)
```
Fixing it seems more tedious...

Reply:
If you know they are the same, you can add an assertion: V.graph.sizevars.guard_equals

Reply:
Yeah, Horace and I discussed about this yesterday. The front-end has done this check, and we're certain these two should be equal.